File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-python/sdkwork_mail_sdk/driver_manager.py
from .provider_activation_catalog import get_mail_provider_activation_by_provider_key
from .provider_catalog import (
    DEFAULT_mail_PROVIDER_KEY,
    MailProviderCatalog,
    get_mail_provider_by_provider_key,
)
from .provider_selection import (
    MailProviderSelection,
    MailProviderSelectionRequest,
    resolve_mail_provider_selection,
)
from .provider_support import (
    MailProviderSupport,
    MailProviderSupportStateRequest,
    create_mail_provider_support_state,
)


class MailDriverManager:
    def resolveSelection(
        self,
        request: MailProviderSelectionRequest | None = None,
        *,
        defaultProviderKey: str = DEFAULT_mail_PROVIDER_KEY,
    ) -> MailProviderSelection:
        return resolve_mail_provider_selection(
            request,
            default_provider_key=defaultProviderKey,
        )

    def describeProviderSupport(self, providerKey: str) -> MailProviderSupport:
        official = get_mail_provider_by_provider_key(providerKey) is not None
        activation = get_mail_provider_activation_by_provider_key(providerKey)

        return create_mail_provider_support_state(
            MailProviderSupportStateRequest(
                providerKey=providerKey,
                builtin=activation.builtin if activation is not None else False,
                official=official,
                registered=activation.runtimeBridge if activation is not None else False,
            )
        )

    def listProviderSupport(self) -> list[MailProviderSupport]:
        return [
            self.describeProviderSupport(entry.providerKey)
            for entry in MailProviderCatalog.entries
        ]

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-rust/src/driver_manager.rs
use crate::provider_activation_catalog::get_mail_provider_activation_by_provider_key;
use crate::provider_catalog::{get_mail_provider_by_provider_key, OFFICIAL_mail_PROVIDERS};
use crate::provider_selection::{
    resolve_mail_provider_selection, MailProviderSelection, MailProviderSelectionRequest,
};
use crate::provider_support::{
    create_mail_provider_support_state, MailProviderSupport, MailProviderSupportStateRequest,
};

pub struct MailDriverManager;

impl MailDriverManager {
    #[allow(non_snake_case)]
    pub fn resolveSelection(
        &self,
        request: &MailProviderSelectionRequest,
        defaultProviderKey: Option<&str>,
    ) -> MailProviderSelection {
        resolve_mail_provider_selection(request, defaultProviderKey)
    }

    #[allow(non_snake_case)]
    pub fn describeProviderSupport(&self, providerKey: &str) -> MailProviderSupport {
        let official = get_mail_provider_by_provider_key(providerKey).is_some();
        let activation = get_mail_provider_activation_by_provider_key(providerKey);

        create_mail_provider_support_state(MailProviderSupportStateRequest {
            providerKey: providerKey.to_string(),
            builtin: activation.is_some_and(|entry| entry.builtin),
            official,
            registered: activation.is_some_and(|entry| entry.runtimeBridge),
        })
    }

    #[allow(non_snake_case)]
    pub fn listProviderSupport(&self) -> Vec<MailProviderSupport> {
        OFFICIAL_mail_PROVIDERS
            .iter()
            .map(|entry| self.describeProviderSupport(entry.providerKey))
            .collect()
    }
}
